plugin/trino-geospatial/src/main/java/io/trino/plugin/geospatial/KdbTreeType.java

```diff
@@ -32,7 +32,10 @@
 
     private KdbTreeType()
     {
-        super(new TypeSignature(NAME), KdbTree.class);
+        // The KDB tree type should be KdbTree but can not be since KdbTree is in
+        // both the plugin class loader and the system class loader.  This was done
```

Comment:
I don't get the implication "class exists in plugin and main => cannot be type representation" yet.
Can you please help me understand this better?

Also. it would be great to test this. If the problem involves classloaders, product tests seem the way to go.

And yet another -- how can we prevent this from happening in the future?

Can TDigest have the same problem? I think it is also loaded by both main and plugin classloader, and is type's stack type.

Reply:
The geo stuff is a pretty big hack, and really different from all other types.  The engine handles interacts directly with KdbTree intances in one of the optimizer.  This mean the KdbTree class must be in the server CL.  This class is not exposed to plugins and therefore when the geo plugin adds the KdbTreeType, it is using a class with the same name, but loaded from a different CL, so it **is** a different class.....  The query in the issue is designed in a way that results in a constant KdbTree bound into an expression.  This constant is instance and the types declared are from the plugin CL, as we expect.  Then we load this generated byte code using the system CL, which also has the KdbTree class.  This step also works.  Finally, when the constant is accessed we attempt to bind the constant from the plugin CL into a place that is using the KdbTree from the system CL, and since they are actually different classes, you get an exception.

This won't affect HyperLogLogType or any other types, because they are using stack types from the JVM or in the SPI.